### PR TITLE
Handling custom type labels in headers from Binary/String message attributes. Fix gh-410 [v2].

### PR DESCRIPTION
**Pull request by voytech**
This is another approach to solve gh-410.
First approach is here: https://github.com/spring-cloud/spring-cloud-aws/pull/633.
Original issue and discussion is here: https://github.com/spring-cloud/spring-cloud-aws/issues/410

### CODE REVIEW

**Comment by maciejwalkowiak** [comment]
I think this is the way to go. Considering we now give an option to just resolve original SQS message as a listener parameter, if someone is interested in more details about the specific message attribute it's still possible.

File: spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java [hunk 1]
```diff
@@ -112,8 +112,8 @@ else if (MessageHeaders.ID.equals(messageAttribute.getKey())) {
 				messageHeaders.put(MessageHeaders.ID,
 						UUID.fromString(messageAttribute.getValue().getStringValue()));
 			}
-			else if (MessageAttributeDataTypes.STRING
-					.equals(messageAttribute.getValue().getDataType())) {
+			else if (messageAttribute.getValue().getDataType()
+					.startsWith(MessageAttributeDataTypes.STRING)) {
```
Comment (eddumelendez) [inline spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java:116]: let's change it to `MessageAttributeDataTypes.STRING.startsWith(messageAttribute.getValue().getDataType())`

File: spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java [hunk 2]
```diff
@@ -122,8 +122,8 @@ else if (messageAttribute.getValue().getDataType()
 				messageHeaders.put(messageAttribute.getKey(),
 						getNumberValue(messageAttribute.getValue()));
 			}
-			else if (MessageAttributeDataTypes.BINARY
-					.equals(messageAttribute.getValue().getDataType())) {
+			else if (messageAttribute.getValue().getDataType()
+					.startsWith(MessageAttributeDataTypes.BINARY)) {
```
Comment (eddumelendez) [inline spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java:126]: let's change it to `MessageAttributeDataTypes. BINARY.startsWith(messageAttribute.getValue().getDataType())`

File: spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java (hunk 1 above)
Comment (voytech) [reply spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java:116]: getDataType may return „String” or string with label which is: „String.<type extensions labels>„

MessageAttributeDataTypes.STRING is always „String” 

If we replace as You suggest we will get:
„String”.startsWith(„String.<Here goes the label>”) which would be rejected, but we want to support everything which starts from „String”

File: spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java (hunk 2 above)
Comment (voytech) [reply spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java:126]: Here is the same as in comment above.

File: spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java (hunk 1 above)
Comment (eddumelendez) [reply spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/core/QueueMessageUtils.java:116]: it was my fault, please ignore these comments. You are right.

**Comment by maciejwalkowiak** [comment]
@voytech purely optional as you already put a lot of effort - could you also update docs?

**Comment by voytech** [comment]
> @voytech purely optional as you already put a lot of effort - could you also update docs?

I have provided some info in docs. You can take a look if this is okey (or sth needs to be changed yet)

**Pull request merged**